### ISSUE TITLE
how to enable additive bias for attention logits using flash attention?

### ISSUE THREAD

**Issue by hiyijian** [HazyResearch/flash-attention#167]
How to enable additive bias for attention logits using flash attention, such as Alibi?

**Comment by tridao** [comment]
This is not supported in the CUDA implementation of FlashAttention.
We have an experimental [implementation](https://github.com/HazyResearch/flash-attention/blob/main/flash_attn/flash_attn_triton.py) in Triton that supports attention bias. You can play with it (it's a self-contained python file) to see if that works for your case.

**Comment by hiyijian** [comment]
Thanks. I put the triton version to Nvidia/NeMo and run FP16 inference on V100  with bloom-560M model(alibi PE.). the result seems reasonable but it is different from pure torch version attention. Any clue about the difference please？

BTW: Is there any plan to implement it using cuda？

**Comment by tridao** [comment]
I'm surprised it (at least the forward pass) worked on V100, I've only tested on A100.

Regarding numerical error, you want to compare 
(FlashAttention in fp16 - standard attention in fp32) vs (standard attention in fp16 - standard attention fp32).
Most of the time FlashAttention gets smaller or around the same numerical error than standard attention.

The CUDA implementation doesn't support attention bias.

**Comment by hiyijian** [comment]
OK. Finally, both training and inference worked like a charm on A100. The difference bettween triton version of flash attention and standard attention disappeared.

Thanks for your excellent effort!

One more question please: Can triton version of flash attention work with BF16 data type?

**Comment by tridao** [comment]
> One more question please: Can triton version of flash attention work with BF16 data type?

Yes, it supports bf16. I've tested it with an older version of Triton (mentioned in the file).

**Comment by rabsher** [comment]
does it work on my local machine or is necessary in need GPU for it to run this model? 

i tried to load a model on my local machine with 15GB of ram but it's stuck model loading because of the size of the model.

There is any way to run it on the local machine? if so please guide me on what the additional parameter needs to pass


`import transformers
import torch
config = transformers.AutoConfig.from_pretrained(
    'mosaicml/mpt-7b',
    trust_remote_code=True
)

config.attn_config['attn_impl'] = 'triton'

model = transformers.AutoModelForCausalLM.from_pretrained(
    'mosaicml/mpt-7b',
    trust_remote_code=True,
    config=config,
    torch_dtype=torch.bfloat16
)
model.to(device='cuda:0')`
